### ISSUE TITLE
Adapting ResNet34 for 3D dataset

### ISSUE THREAD

**Issue by willi3by** [ageron/handson-ml2#48]
Hi, 

First, thank you very much for your work, it is fantastic and I have enjoyed working through the book very much. 

I am trying to implement the ResNet34 from scratch that you show in the book on a 3D dataset but have ran into an error message. When executing the for loop that adds the residual layers to the model, on the 5th iteration, I get the error message:

"Dimensions must be equal, but are 128 and 64 for 'residual_unit_35/conv3d_243/convolution' (op: 'Conv3D') with input shapes: [?,12,14,12,128], [3,3,3,64,128]."

Here is the current code I am using to build the model:

```python
class ResidualUnit(keras.layers.Layer):

    def __init__(self, filters, strides=1, activation="relu", **kwargs):
        super().__init__(**kwargs)
        self.activation = keras.activations.get(activation)
        self.main_layers = [
            keras.layers.Conv3D(filters, (3,3,3), strides=strides,
                                padding="same", use_bias=False),
            keras.layers.BatchNormalization(),
            self.activation,
            keras.layers.Conv3D(filters, (3,3,3), strides=1,
                                padding="same", use_bias=False),
            keras.layers.BatchNormalization()]
        self.skip_layers = []
        if strides > 1:
            self.skip_layers = [
                keras.layers.Conv3D(filters, (1,1,1), strides=strides,
                                    padding="same", use_bias=False),
                keras.layers.BatchNormalization()]

    def call(self, inputs):
        Z = inputs
        for layer in self.main_layers:
            Z = layer(Z)
        skip_Z = inputs
        for layer in self.skip_layers:
            skip_Z = layer(skip_Z)
        return self.activation(Z + skip_Z)

model = keras.models.Sequential()
model.add(keras.layers.Conv3D(64, (7,7,7), strides=2, padding="same",
                              use_bias=False, input_shape=[91,109,91,1]))
model.add(keras.layers.BatchNormalization())
model.add(keras.layers.Activation("relu"))
model.add(keras.layers.MaxPool3D(pool_size=3, strides=2, padding="same"))
prev_filters = 64
for filters in [64] * 3 + [128] * 4 + [256] * 6 + [512] * 3:
    strides = 1 if filters == prev_filters else 2
    model.add(ResidualUnit(filters, strides=strides))
    prev_filters = filters
    print(filters)
model.add(keras.layers.GlobalAveragePooling3D())
model.add(keras.layers.Flatten())
model.add(keras.layers.Dense(2, activation="sigmoid"))
```
Do you have any advice for resolving this?

Thank you,
Brady

**Comment by ageron** [comment]
Hi @willi3by,

This looks really cool, and I see nothing wrong with your code, except perhaps the fact that the final `Dense` layer should have 1 unit if you're trying to do binary classification: it will output the probability of the positive class. If you're trying to do two binary classifications simultaneously (e.g., answering two separate binary questions such as "is there a tumor in the 3D image?" and "is there a haemorrhage in the 3D image?").

I tried running the code in [Colab](https://colab.research.google.com), and it works just fine, you can try it now. Just add this before your code:

```python
%tensorflow_version 2
import tensorflow as tf
from tensorflow import keras
```

and add this just after your code:

```python
model.summary()
```

You will get this output:

```
Model: "sequential"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
=================================================================
conv3d (Conv3D)              (None, 46, 55, 46, 64)    21952     
_________________________________________________________________
batch_normalization (BatchNo (None, 46, 55, 46, 64)    256       
_________________________________________________________________
activation (Activation)      (None, 46, 55, 46, 64)    0         
_________________________________________________________________
max_pooling3d (MaxPooling3D) (None, 23, 28, 23, 64)    0         
_________________________________________________________________
residual_unit (ResidualUnit) (None, 23, 28, 23, 64)    221696    
_________________________________________________________________
residual_unit_1 (ResidualUni (None, 23, 28, 23, 64)    221696    
_________________________________________________________________
residual_unit_2 (ResidualUni (None, 23, 28, 23, 64)    221696    
_________________________________________________________________
residual_unit_3 (ResidualUni (None, 12, 14, 12, 128)   673280    
_________________________________________________________________
residual_unit_4 (ResidualUni (None, 12, 14, 12, 128)   885760    
_________________________________________________________________
residual_unit_5 (ResidualUni (None, 12, 14, 12, 128)   885760    
_________________________________________________________________
residual_unit_6 (ResidualUni (None, 12, 14, 12, 128)   885760    
_________________________________________________________________
residual_unit_7 (ResidualUni (None, 6, 7, 6, 256)      2690048   
_________________________________________________________________
residual_unit_8 (ResidualUni (None, 6, 7, 6, 256)      3540992   
_________________________________________________________________
residual_unit_9 (ResidualUni (None, 6, 7, 6, 256)      3540992   
_________________________________________________________________
residual_unit_10 (ResidualUn (None, 6, 7, 6, 256)      3540992   
_________________________________________________________________
residual_unit_11 (ResidualUn (None, 6, 7, 6, 256)      3540992   
_________________________________________________________________
residual_unit_12 (ResidualUn (None, 6, 7, 6, 256)      3540992   
_________________________________________________________________
residual_unit_13 (ResidualUn (None, 3, 4, 3, 512)      10754048  
_________________________________________________________________
residual_unit_14 (ResidualUn (None, 3, 4, 3, 512)      14159872  
_________________________________________________________________
residual_unit_15 (ResidualUn (None, 3, 4, 3, 512)      14159872  
_________________________________________________________________
global_average_pooling3d (Gl (None, 512)               0         
_________________________________________________________________
flatten (Flatten)            (None, 512)               0         
_________________________________________________________________
dense (Dense)                (None, 1)                 513       
=================================================================
Total params: 63,487,169
Trainable params: 63,470,145
Non-trainable params: 17,024
_________________________________________________________________
```

Just curious: what are you using this for? Is it a medical application?

I'm not sure why this could would fail on your machine. Perhaps you don't have the latest TensorFlow version installed? Or maybe you're working in a Jupyter notebook and there was some previous state causing problem? In this case try restarting the Jupyter kernel.

I hope this helps.

**Comment by willi3by** [comment]
Very interesting, when I tried to use ```from tensorflow import keras```  instead of ```import keras```, everything works as expected. Thank you very much! 

This is indeed for medical applications. I am a neuroimaging researcher so all of our data is in 3 or 4 dimensions (space+time). Thanks for catching the typo in the Dense layer, the current application is for binary classification but future models will be aimed at 2 or more simultaneous binary classifications. 

Since we are in the discussion, I was wondering if you had advice on a couple of other potential applications. 
1. For the simultaneous binary classification, say I have 4 binary groups, would I just use 4 units in the Dense layer with Sigmoid activation? 
2. Do you have a recommendation for combining multiple 3D feature sets in a single network? Right now the two options I can come up with are: 1) running each 3D dataset through its own ResNet, combining the average pooling layer from each modality into a "final" feature set, then using this for the prediction; this would allow backtracking to see which modality was most predictive, or 2) stacking the images into a 4D tensor and using this as the input. 

Thank you very much for your time and, again, for your terrific book!

Brady

**Comment by ageron** [comment]
Hi @willi3by ,

I'm glad this solved the problem! :) I'm super excited whenever I chat with a reader who uses my book for an awesome application, and I can't think of better applications than medical applications, I truly hope your system will be super successful.

Regarding the import: `import keras` imports the "multi-backend keras" implementation of the keras API (it's what you get when you run `pip install keras`). Unless you plan to use another backend than TensorFlow (for example Theano), you can safely run `pip uninstall keras`, you won't need it. When you run `from tensorflow import keras`, you're using the implementation of the Keras API that comes bundled with TensorFlow. It rocks.

To answer your question 1: yes, you just need to have 4 units in the `Dense` layer, use the `sigmoid` activation function, and use the `binary_crossentropy` loss. This gives you 4 binary outputs. If instead you have 4 mutually exclusive classes, then you need the `softmax` activation function and the `categorical_crossentropy` loss (or `sparse_categorical_crossentropy` loss if the labels are represented as class indices rather than class probability).
 
Regarding your second question, judging from the `input_shape` in your code, which is `[91,109,91,1]`, it seems that your inputs only have very few channels (just one in this case), so I would just stack the multiple 3D feature sets along the channel dimension and use the 3D ResNet you built (I think that's what you meant in your second option, but I wasn't sure whether you meant using 4D convolutions, which would be unnecessary).

```python
import numpy as np

X1 = np.arange(24).reshape(2,3,4,1) + 100
X2 = np.arange(24).reshape(2,3,4,1) + 200
X3 = np.arange(24).reshape(2,3,4,1) + 300

X = np.concatenate([X1, X2, X3], axis=-1)

print("X1 =")
print(X1)

print("X =")
print(X)
```

Displays this:

```
X1 =
[[[[100]
   [101]
   [102]
   [103]]

  [[104]
   [105]
   [106]
   [107]]

  [[108]
   [109]
   [110]
   [111]]]


 [[[112]
   [113]
   [114]
   [115]]

  [[116]
   [117]
   [118]
   [119]]

  [[120]
   [121]
   [122]
   [123]]]]

X =
[[[[100 200 300]
   [101 201 301]
   [102 202 302]
   [103 203 303]]

  [[104 204 304]
   [105 205 305]
   [106 206 306]
   [107 207 307]]

  [[108 208 308]
   [109 209 309]
   [110 210 310]
   [111 211 311]]]


 [[[112 212 312]
   [113 213 313]
   [114 214 314]
   [115 215 315]]

  [[116 216 316]
   [117 217 317]
   [118 218 318]
   [119 219 319]]

  [[120 220 320]
   [121 221 321]
   [122 222 322]
   [123 223 323]]]]
```

Your first option is also worth trying. I suspect it will not perform as well, but it's always hard to guess.

I hope this helps.

**Comment by willi3by** [comment]
Thank you, that is very kind! 

That is good to know, I will definitely do that to prevent this kind of issue in the future. 

That is very useful, I am attempting to calculate a composite score from 4 different measures (all scored 0 or 1). So I am trying to think of the best way to go about it. Im not sure if I should try to predict each class individually and add the results from the Dense layer or use the composite score as ordinal labels (0-4). But your advice definitely gets me on the right track! 

This is super interesting! I never thought of this approach, I was referring to 4D convolutions but I think adding by channel is a great idea. All of the images we use are grayscale so they only have a single channel. 

Thanks again for your help,
Brady
